### PR TITLE
fix(batch): Add casting for the equality columns of the join predicate in Lookup Join

### LINKED ISSUE #4044: Lookup Join throws an error when equality columns of join predicate have different data types
**Describe the bug**
When the equality columns of a join predicate in Lookup Join have different data types, the executor will throw an error instead of attempting to cast. This means even data types that are compatible with each other (i.e. int32 and int64) will still cause an error.

**To Reproduce**
```
set query_mode = local;
set rw_batch_enable_lookup_join = true;
create table t1 (v1 bigint, v2 int);
create table t2 (v1 int, v2 int);
create materialized view t3 as select v1, count(*) as v2 from t2 group by v1;
insert into t1 values (1, 3);
insert into t2 values (1, 4);
select * from t1 join t3 on t1.v1 = t3.v1;
```

Actual behavior: A `QueryError` is thrown and an error message is displayed.

**Expected behavior**
The query runs without error.
```
 v1 | v2 | v1 | v2 
----+----+----+----
  1 |  3 |  1 |  1
(1 row)
```

**Additional context**
Might be related: #3478

### PR DESCRIPTION
**Pull request by Graphcalibur**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
    - In Lookup Join, when converting the datums in the build side key columns, the code now attempts to cast it into the data type on the probe side key column instead of throwing an error whenever the data types are different

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
Fixes #4044

### CODE REVIEW

File: src/batch/src/executor/join/lookup_join.rs [hunk 1]
```diff
@@ -162,9 +165,30 @@ impl<C: BatchTaskContext> ProbeSideSourceBuilder for ProbeSideSource<C> {
...
+                    Type::Cast,
+                    probe_type.clone(),
+                    Box::new(LiteralExpression::new(
+                        build_type.clone(),
+                        Some(scalar_impl.clone()),
+                    )),
+                )?;
+
+                let datum = cast_expr.eval_row(&Row(vec![]))?;
+                datum.unwrap()
```
Comment (lmatz) [inline src/batch/src/executor/join/lookup_join.rs:187]: Does it mean that `datum` cannot be `None`?

**Review by lmatz — APPROVED** [review]

File: src/batch/src/executor/join/lookup_join.rs (hunk 1 above)
Comment (Graphcalibur) [reply src/batch/src/executor/join/lookup_join.rs:187]: Yep

Comment (Graphcalibur) [reply src/batch/src/executor/join/lookup_join.rs:187]: Actually a panic occurs if `null` is inserted into the table but it's an issue caused by the other parts of the code. I think I can fix this behavior in a different PR.

Comment (lmatz) [reply src/batch/src/executor/join/lookup_join.rs:187]: Can add the `mergify/can-merge` to let this PR automatically merge with the main branch.

**Pull request merged**